### ISSUE TITLE
SLIP-19 ownership proofs

### ISSUE THREAD

**Issue by matejcik** [spesmilo/electrum#8325]
follow-up to the previous issue https://github.com/spesmilo/electrum/issues/8324 -- Trezor will only accept external inputs that are pre-signed, or have a [SLIP-19 ownership proof](https://github.com/satoshilabs/slips/blob/master/slip-0019.md) attached.

This means that it's not possible to do manual coinjoin between two Electrum+Trezor wallets, even if #8324 is fixed, because both sides will refuse to be the first to sign.

A solution is adding the option to export the unsigned transaction with ownership proofs attached. Then Electrum could even generate ownership proofs for its hot wallets for perfect(tm) compatibility .)

I can advise on how exactly to get the ownership proof from Trezor, maybe even make a PR if I can find the time to do it.

I would however need advice on where exactly to plug this into the signing flow, because it requires active confirmation on Trezor.
Add "Export -> With ownership proof" option which triggers the flow? A separate button on the transaction screen? Do it automatically when you click Pay? Something else?

**Comment by SomberNight** [comment]
> I can advise on how exactly to get the ownership proof from Trezor

Please do. I am playing around with this (https://github.com/SomberNight/electrum/tree/202402_slip19_trezor) and am a bit confused about the API/requirements.

Does this look ok?
https://github.com/SomberNight/electrum/commit/6ddb81e7c2d422e4b10dfea4080703ad76bfcbc6#diff-2881a772e94070a1569d653bbbe84089a71ea96168cfc3f8edd5206862620b5fR385-R390
I am getting `DataError: Invalid witness.` from the firmware (trying with model T, all p2wpkh inputs)
I am setting `txinputtype.ownership_proof` to `proofBody || proofSignature`, is that correct?

**Comment by matejcik** [comment]
I believe the signature is already part of the returned `proof`, and the reason for a separate field is just convenience for callers who need it separately. (this would explain the "invalid witness" error because there is now trailing garbage in the witness data)

for reference, here is the code that generates the proof:
https://github.com/trezor/trezor-firmware/blob/grdddj/fw_translations/core/src/apps/bitcoin/ownership.py#L32-L78
the function [write_bip322_signature_proof](https://github.com/trezor/trezor-firmware/blob/main/core/src/apps/bitcoin/scripts.py#L611) accepts signature as an argument and writes it out

**Comment by matejcik** [comment]
otherwise the code looks good, LMK if it still doesn't work and I'll see what's going on in the deep layers

**Comment by SomberNight** [comment]
>> I am setting `txinputtype.ownership_proof` to `proofBody || proofSignature`, is that correct?
>
> I believe the signature is already part of the returned `proof`

Hmm but if I use just that (https://github.com/spesmilo/electrum/commit/c8ab78f43f6f11a3001fb6473c3c14c09efc37ea), then I get `DataError: Invalid external input` from the firmware.

**Comment by matejcik** [comment]
ahahah but you are doing it on the same Trezor seed, aren't you?
The "ownership proof" actually proves _non-ownership_. If the input belongs to the same seed, the check will fail.

Try using two different passphrases for the inputs.

(To get two inputs from the same seed signed, you can't use external inputs, you would have to request signing of both in the same operation. Otherwise it's essentially the exact attack that this feature protects against.)

**Comment by SomberNight** [comment]
Oh. :D
I see.

**Comment by SomberNight** [comment]
Re multisig, looking at the PSBT fields and the [workflow in the spec](https://github.com/satoshilabs/slips/blob/b6b8ade46333e782eeded364df9cbcfd113ce77d/slip-0019.md#multi-signature-scriptpubkeys), do I understand correctly that the idea is that the coordinator (e.g. Electrum) is connected to all signing devices?

So let's consider k-of-n multisig. When adding the ownership proofs, the Electrum should
- run `client.get_ownership_id()` for each of the `n` clients, for each txin/address
  - this step could be cached and e.g. done when generating addresses, e.g. rolling the gap limit forward (and stored in the wallet file)
    - still, it requires all `n` devices (though it could be done opportunistically and always done for whatever device is connected)
  - no explicit support for moving partial data around in a PSBT
- run `client.get_ownership_proof()` for a quorum of  `k` clients, for each txin/address
  - no explicit support for moving partial data around in a PSBT

So basically without persistence for step1, to export a tx, all `n` devices need to be connected, right?

**Comment by matejcik** [comment]
I'm gonna tag in @andrewkozlik for this one
